### ISSUE TITLE
Improving T5 Docs

### ISSUE THREAD

**Issue by alexcoca** [huggingface/transformers#16614]
### Who can help
@NielsRogge @patrickvonplaten @sgugger 

Documentation: @sgugger

## Information

Model I am using : T5ForConditionalGeneration

The problem arises when using:
* my own modified scripts

The tasks I am working on is:
*  my own task or dataset

## To reproduce

#[13240 ](https://github.com/huggingface/transformers/pull/13240)  is a really nice PR that adds a lot of clarity to the documentation. However, in the examples provided we read

```python
tokenizer.pad_token = tokenizer.eos_token  # to avoid an error
```
and I feel more information could be given to explain users why this is necessary. I am currently attempting to do batched decoding with T5 and observing very strange outputs, and therefore I'm keen to understand if this is a problem. Very soon I will test to understand whether the strange behaviour is due to batching or not, but it would be great to enhance the docs to explain the error that would occur and why.

## Expected behavior

One or two extra sentences to explain why we left-pad encoder input sequences with <EOS> when doing batched decoding for T5.

**Comment by NielsRogge** [comment]
Hi,

Thanks for your question! To be honest it wasn't clear for me neither, I guess it's set as otherwise it might complain that no padding token is set.

I took that snippet from this PR: https://github.com/huggingface/transformers/pull/7552#event-3876130796.

It includes the comment:
```
# when generating, we will use the logits of right-most token to predict the next token
# so the padding should be on the left
```
However, that was for a decoder-only model (GPT-2). Not sure whether the same is required for an encoder-decoder one like T5. Maybe @patrickvonplaten can clarify here.

**Comment by alexcoca** [comment]
Thanks @NielsRogge, this would be very helpful indeed. I did the following checks this morning:

1. Run my original, non-batched decoder
2. Run batched decoding, implemented following the docs

The results are vastly different. 

For 1 we get:

```json
{
    "1_00000": {
        "0": {
            "utterance": "hi, could you get me a restaurant booking on the 8th please?",
            "Restaurants_2": {
                "predicted_str": " [states] 10:the 8th [intents] i1 [req_slots] <EOS>"
            }
        },
        "1": {
            "utterance": "could you get me a reservation at p.f. chang's in corte madera at afternoon 12?",
            "Restaurants_2": {
                "predicted_str": " [states] 0:corte madera 2:the 8th 9:p.f. chang's 10:afternoon 12 [intents] i1 [req_slots] <EOS>"
            }
}
```

For 2 we get:

```json
{
    "1_00000": {
        "0": {
            "utterance": "hi, could you get me a restaurant booking on the 8th please?",
            "Restaurants_2": {
                "predicted_str": " [states] 10:the 8th [intents] i1 [req_slots] i1 [req_slots] i1 [req_slots] i1 [req_slots] i1 [req_slots] i1 [req_slots] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents]  <EOS>"
            }
        },
        "1": {
            "utterance": "could you get me a reservation at p.f. chang's in corte madera at afternoon 12?",
            "Restaurants_2": {
                "predicted_str": " [states] 0:corte madera 2:the 8th 9:p.f. chang's 10:afternoon 12 [intents] i1 [req_slots] i1 [req_slots] i1 [req_slots] i1 [req_slots] i1 [req_slots] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] [intents] 0 [intents] i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i1 i <EOS>"
            }
        },
}
```

Therefore the must be a subtle bug in my implementation of the the batched decoding. The results above are obtained with `batch_size=5` buy setting `batch_size=1` gives identical results.

I'll try to figure out why this occurs. I should state the I can make the code/checkpoints to replicate the above issues available to you for debugging. I'll keep you posted.

**Comment by alexcoca** [comment]
Ok, I debugged my code and the preliminary test passed. The change has been to change the `generate` API call from 


```python
output_seqs = model.generate(
    input_ids=input_ids.to(DEVICE),
    attention_mask=attention_mask.to(DEVICE),
    max_length=args.decoder_max_seq_len,
    use_cache=True,
)
```

to 

```python
output_seqs = model.generate(
    input_ids=input_ids.to(DEVICE),
    attention_mask=attention_mask.to(DEVICE),
    max_length=args.decoder_max_seq_len,
    bos_token_id=tokenizer.bos_token_id,
    eos_token_id=tokenizer.eos_token_id,
    pad_token_id=tokenizer.pad_token_id,
    use_cache=True,
)
```

I'm not sure if this is an omission in the docs or why this is a fix. An explanation would be appreciated! Maybe something note worthy is that the last id in the `input_ids` is the `<EOS>`  token ID.

**Comment by patrickvonplaten** [comment]
Good point @alexcoca ,

@NielsRogge @alexcoca  Yeah this looks like it was a bad copy-paste from GPT2.

Should be corrected here: https://github.com/huggingface/transformers/pull/16646

**Comment by patrickvonplaten** [comment]
I could be related to https://github.com/huggingface/transformers/issues/14859#issuecomment-1015389797 actually
